### DISCUSSION TITLE
FTXUI + KiTTY terminal graphics protocol?

### DISCUSSION THREAD

**Discussion by spinicist** [ArthurSonzogni/FTXUI#824] [Q&A]
Hello, has anyone tried combining FTXUI with the KiTTY terminal graphics protocol? I have a basic terminal viewer for a specialized image format working with the KiTTY protocol, I think adding FTXUI to this could make something pretty awesome. But I don't know how to combine the two (i.e. draw an image using the KiTTY protocol in a specific FTXUI component). Thanks.

**Comment by ArthurSonzogni** [comment]
Hello @spinicist.
You probably can use both. I don't see any potential problems.
Please let us know what you got ;-)

**Reply by spinicist** [reply]
Can you give me a hint for a good place to start trying this? I assume I want to subclass Canvas or a similar class. I'm struggling to get my head round which member function of each widget is responsible for "drawing".

**Reply by ArthurSonzogni** [reply]
I use Kitty, but I don't know the kitty extensions.

As long as you only need to send data to kitty, this should be as simple as writting to stdout using the kitty [documentation](https://sw.kovidgoyal.net/kitty/graphics-protocol/#the-graphics-escape-code).
```cpp
std::cout << "<ESC>_G<control data>;<payload><ESC>";
```
This could stay outside of FTXUI.

We know that:
- FTXUI is always writting data from the main thread. So do should do the same.
- FTXUI is always waiting for data from a dedicate thread.

In FTXUI:
- A `ftxui::Components` are creating the `ftxui::Element` to draw a frame.
- A `ftxui::Element` is writting on the `ftxui::Screen`.
- A `ftxui::Screen` is writting on stdout. See [code](https://github.com/ArthurSonzogni/FTXUI/blob/main/src/ftxui/screen/screen.cpp#L449C6-L449C20).

Note: Ideally this would be implemented as a Component or element outside of FTXUI so that developers can opt-in using it. I'd be rather reluctant to add it to FTXUI.

**Reply by spinicist** [reply]
Indeed, the KiTTY graphics protocol is fairly straightforward. My existing code that prints direct to the terminal is here: https://github.com/spinicist/riesling/blob/main/src/cmd/montage.cpp#L102

However the aim of swapping to FTXUI would be too have some kind of bitmap drawing class that would rescale the image to fit within the bounds of the widget and all that good stuff. That's why I think I need to integrate it into the FTXUI class hierarchy.

**Reply by spinicist** [reply]
I completely understand about keeping it opt-in / outside the FTXUI repo. Let's see what I can get working first!

**Reply by spinicist** [reply]
Big thanks for pointing to the `ftxui::Screen` writing code. Sadly, I think this design will make it hard to integrate KiTTY graphics. I'm not criticising FTXUI, in fact I think it's a very sensible design. But I don't think it plays well with the way the KiTTY protocol works. I don't see what the `Element` would write onto the `Screen` to then be written to `stdout`, as it needs to be in escape codes. I will have to think more.
